src/renderer/canvas/canvas-renderer.ts

```diff
@@ -195,6 +256,23 @@ export class CanvasRenderer<N extends INodeBase, E extends IEdgeBase> extends Em
     }
   }
 
+  private _resize() {
+    const dpr = this._settings.devicePixelRatio || window.devicePixelRatio;
```

Comment:
Nitpicking, but it would be nice to add a check for `window.devicePixelRatio` being null or undefined, just in case of older browsers that do not support this. I think using `|| 1` will do the job.

Reply:
Not nitpicking, but it's a valid bug. Great catch.